Penrose Tiling

ceci-fidler opened the pull request:
First version of code to create the Penrose Tilings

ceci-fidler committed:
bump

ceci-fidler committed:
Thick Rhombus

Drawing a thick rhombus, with a centroid at (0,0), to the screen via the FirstTile method

ceci-fidler committed:
Thick and Thin Rhombus

Drawing both a thick and thin rhombus on the screen; Coordinates for each of the rhombi are based on a centroid at (0,0); Thick rhombus is drawn at (0,0) on screen; Thin rhombus is drawn at (0,-1) on screen

ceci-fidler committed:
First Pass - Generate Line, Get Intersection, Draw Tile

Established a basic method that generates a line as a member in one of 5 families based on the indexing (family #, member #) system; Established a basic method that finds an intersection between 2 members in their respective families; For now, that intersection point is returned to the firstTile method, where a basic square will be drawn at that point; No specified orientation or thick/thin rhomb is determined at an intersection yet; Reverted to square polygon for this first pass due to initial logic error and debugging via console output

ceci-fidler committed:
2/14 Updates and rhombusType Method

February 14th meeting updates, including revising the defineLine method to avoid overlapping intersections in the pentagrid; Defined a read-only array of offset values c for each family that is passed into the constructor; Incorporated gridLineIndex type where it should have been utilized before; rhombusType method places the correct rhombus at the specified intersection via combinatorial pattern in the firstTile method

ceci-fidler committed:
Work on 2/21 through 2/28 - Pentagrid, Rhombs at Location with Rotation!

2/21 - Wrote a method to scale an affine polygon (affine-polygon.ts); This is so we could draw tiny rhomb tiles at each intersection; Drew the pentagrid in overwrite Draw method using the intersection of an arbitrary Line with a Circle to draw very long, finite line segment; Drew each tile out in overwrite Generate at each intersection point - no rotation method to be done as of yet, just minor adjustments to other methods (type changes, etc.) to get things working ; ; ; 2/28 - Figured out the angle rotations at all 10 unique intersections and wrote the rhombusRotation method; Combinatorial formula is not found yet, but this method works (a good start); Also commented out this method and Generate first and played around with hardcoded angles in the firstTile method; Also change the color of the thick rhomb to orange and the thin rhomb to cyan; Played around with how many rhomb w/ rotation we could place at intersection points by changing the integer passed into Generate via tbComponent (this controls the depth parameter in the override Generate)

ceci-fidler committed:
Getting a Normal Vector to Each Rhomb Side

The goal is to pass into the findNextIntersection method: a normal vector corresponding to each side of each respective rhomb in order to be able to determine direction when we want to move to find the next intersection point; We successfully have done this by writing a sideNormal method in affine-Polygon, then passing it in as the third argument via the findNextIntersection method; Verified through debugging console output statements that normal vectors in correct direction facing out from the rhomb were being calculated

ceci-fidler committed:
Identifying Which Line We are "On" Based on Normal Vector Orientation

In findNextIntersection, I was able to orient the normal vector that I passed in based on the rotation of the given rhomb at the intersection, called the orientationDirection. Then, I added the intersection point vector and the orientationDirection vector to get a translated point, which would live on either of the two gridline indices. From there, I had trouble using the containPoint method, as it was being finnicky and not returning the correct answers though the logic was sound. Also, the intersectLine method returns an intersection point that is “incorrect” where it is scaled by a factor of -1. This causes an issue, but it seems that the codebase is working in a sense around this error. Trying to change it to the correct formula, the containsPoint method causes a ‘Bad Intersection’ error to be thrown – very weird – it seems that method is incredibly sensitive.

The way in which Line objects are created is indeed correct. So I wrote a new method called lineContainsTranslatedPoint, which takes in a Line object and a 2D vector and outputs a Boolean. We calculate LHS = a*x + b*y and then RHS = -c (accounts for the weird -1 factor on our intersection point that we have to deal with) . Then, it takes abs(LHS-RHS) < small EPSILON. I first just equated LHS ==RHS, but I got the wrong Booleans returned due to miniscule differences in RHS & LHS due to computational round-off error.

In a case structure, I passed in our lines using defineLine with each one of our gridline indices and then the intersection point; Via debugging output, as adjacent tile goes around each side, it is able to identify each line we are supposed to move along based on it finding the translatedPoint on that line via the normal vector; Everything looks correct

Now we know how to classify which GLI as the one we are “on” and then the one that is just “crossing”, which begins the stages for iterating through the other 4 familes, finding the closest intersections for each one; Then, from those, sorting the one out with least distance to find the next intersection we move to to generate a tile

ceci-fidler committed:
Running into Edge Case Issues

It seems like the orientation of line members can switch whether we need to take the floor or ceiling depending on our strategy of finding distance in terms of L's; The members of first family of lines (vertical red) are indexed like this across the screen like this: ... 3 2 1 0 -1 -2 -3 ...

ceci-fidler committed:
3/5 Updates

Very close to the Penrose; Throwing error in adjacentTile where the normal vector is not being found

ceci-fidler committed:
3/7 debug solved

family 2 and 3 case -> should be 4pi/5, not 3pi/10

ceci-fidler committed:
Finishing the Penrose Tiling

Lael commented on angular/TilingBilliards/src/math/geometry/line.ts:
```diff
@@ -55,8 +55,12 @@ export class Line {
         const solution = new Complex(
-            -other.b * this.c + this.b * other.c,
-            other.a * this.c - this.a * other.c,
+          -other.b * this.c + this.b * other.c,
+            // -other.b * this.c + this.b * other.c, - incorrect numerator to find x coordinate
+            // other.b * this.c - this.b * other.c,
+          other.a * this.c - this.a * other.c,
+            // other.a * this.c - this.a * other.c, - incorrect numerator to find y coordinate
+            // this.a * other.c - other.a * this.c,
```
Please revert

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/math/geometry/line.ts:
```diff
@@ -79,4 +83,11 @@ export class Line {
\ No newline at end of file
+
+    containsVector(vector: Vector2): boolean {
+
+       let d = vector.dot(new Vector2(this.a, this.b));
+
+      return closeEnough(d,0);
```
Please remove extra whitespace

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/tile-billiards/affine-polygon.ts:
```diff
@@ -10,92 +10,116 @@ export interface AffinePolygonRayCollision {
+  // return normal vector to the side of rhomb
+
+  sideNormal(sideIndex: number): Vector2 {
+
```
Remove extra newline

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/tile-billiards/affine-polygon.ts:
```diff
@@ -10,92 +10,116 @@ export interface AffinePolygonRayCollision {
+    let vector2 = this.vertices[(sideIndex+1) % this.n];
+
+    return vector1.clone().add(vector2).multiplyScalar(0.5);
+
```
Remove extra newline

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tile.ts:
```diff
@@ -0,0 +1,22 @@
+
+    this.gridLineIndex1 = gridLineIndex1;
+    this.gridLineIndex2 = gridLineIndex2;
+
+  }
+
```
Remove extra newlines

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/tile-billiards/tiling-billiards-component.ts:
```diff
@@ -62,7 +66,12 @@ export class TilingBilliardsComponent extends ThreeDemoComponent implements OnDe
+       this.tiling = new PenroseTiling([this.c1,this.c2,this.c3,this.c4,this.c5]);
+       //this.tiling.generate(1);
+      this.tiling.generate(this.depth);
+      /*
```
Please remove commented code

ceci-fidler replied:
Done

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+export class PenroseTiling extends AffinePolygonalTiling<PenroseTile> {
+
+  constructor(readonly c: number[]) {
+    // TODO create indices of rhombi (both thick and thin) to draw on the screen - DONE
+    // polygon: AffinePolygon.regular(4,1) - used to confirm a polygon drawn to the screen
+
```
Remove comments if they are no longer relevant and do not commit commented-out code

ceci-fidler replied:
Done - all comments like this should be removed now

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
General style comment (i.e., not very important): sometimes blank lines in code can help readability, but too many can be as hard to read as not enough. I would like to have one blank line between methods and between different conceptual phases of the same method, but never just after a { or just before a }.

ceci-fidler replied:
Got it - wherever possible on this commit I spaced the methods by 1 blank line, as well as instituted some in methods between phases, but let me know what you think

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+    ])
+  }
+
+  /*
+  Going to write a method that defines a line in a family specified by the family and member tags
+   */
```
Here and throughout, please write comments that describe the current functionality of non-trivial methods.

ceci-fidler replied:
Done - comments for non-trivial methods have been updated as such

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+  rhombusType(l1: GridLineIndex, l2: GridLineIndex): number {
+
+    const modValue = Math.abs(l2.family - l1.family) % 5;
+    if (modValue == 2 || modValue == 3)
+      return 1;
+    else
+      return 0;
```
This can be shortened to one line with a ternary operation:
`return (modValue == 2 || modValue == 3) ? 1 : 0;`

ceci-fidler replied:
Done - change to method was made

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+
+    let contains  = false;
+
+    let RHS = -testLine.c     // negate c to adjust for negative intersection output
+    let LHS = testLine.a*translatedPoint.x + testLine.b*translatedPoint.y;
```
Prefer lowercase names for local variables

ceci-fidler replied:
Done - these lines are gone

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+    let RHS = -testLine.c     // negate c to adjust for negative intersection output
+    let LHS = testLine.a*translatedPoint.x + testLine.b*translatedPoint.y;
+
+    console.log(`LHS: ${LHS}, RHS : ${RHS}`); // debug
```
Please do not commit log statements.

ceci-fidler replied:
Done, this method no longer exists

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+
+    console.log(`LHS: ${LHS}, RHS : ${RHS}`); // debug
+
+    if (Math.abs(LHS-RHS) < 1e-6) {   // can have miniscule differences due to round off error in computations
+
+       contains = true;
+    }
+
+    return contains;
```
`return closeEnough(LHS, RHS);`

ceci-fidler replied:
Done, this method no longer exists

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+    return contains;
+  }
+
+  findDistanceInTermsOfL(distance: number, L: number): number {
+
+    return distance/L;
+  }
```
Can this be inlined? Meaning, do we need a method for this?

ceci-fidler replied:
method not used so it has been removed

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+
+  findMemberIndex(tileIntersection: Vector2, direction: Vector2,
+                  familyIndex: number , lineOn: GridLineIndex, lineCrossing: GridLineIndex): number {
+
+    //debug
```
There are blank lines after almost every line of code in this method. I find it much clearer to use blank lines to separate more conceptually distinct pieces of work in a method.

ceci-fidler replied:
Done, I believe this method is either gone entirely or much different

Lael commented on angular/TilingBilliards/src/tile-billiards/penrose-tiling.ts:
```diff
@@ -0,0 +1,429 @@
+      next = this.findNextIntersection(t.gridLineIndex2, t.gridLineIndex1, normalVector);
+      lineOn = t.gridLineIndex2;
+    } else {
+      debugger;
```
Please remove debugger statements from committed code.

ceci-fidler replied:
Done

Lael reviewed (changes requested):
Hi Cecilia! Since we worked through the code together, pretty much all my comments are aesthetic in nature, but I think it would be good to take a few minutes at this stage to step through and tidy up the code. Now that things are working, it's important to make sure the code is as readable as possible, meaning that
- variable and method names are descriptive with a level of detail fitting their scope (i.e., a loop index can be `i` while a parameter name should be a bit more verbose and a public method name should leave no doubt as to what it does)
- comments should document what the code does any time it's not obvious
- spacing should follow the standard conventions
- tslint should be happy (we can work through this in person).

Since this codebase forms the core of your thesis, let's have it looking its best!

ceci-fidler committed:
Unneeded Methods and Comments

Deleting unneeded methods and commenting

ceci-fidler committed:
Getting Billiard Path Working

Trying to get the billiard infrastructure working with a boolean flag that controls behavior (using reflection vs Snell's law); Running into problem right now where the collision point array will only consist of the first collision point and we can't draw the full billiard path

ceci-fidler committed:
Addressing Aesthetic Concerns from Pull Request

Addressing the aesthetic concerns regarding spacing, etc. in the first PR request before merging; Still need to review/update variable names and review/update comments  for non-trivial methods

ceci-fidler committed:
First Round Cleanup and Editing

Went through and cleaned up rogue comments and spacing left over from building process this semester; Simplified methods like rhombusRotation with combinatorial formula and inlining; Reviewed some variable names/commented less-obvious methods of PenroseTiling; Integrated choosing refractive indices in the GUI alongside a checkbox for Snell's Law (starts out at 1 and -1 or select new indices before checking box - if you want to change indices while Snell's Law is checked, just change billiard iterations for new refract. indices to implement); Changed color of billiardPath to blue as well as added a small red circle at the start of the billiardPath (billiardStart) for ease of seeing open and closed orbits

Lael commented on angular/TilingBilliards/src/tile-billiards/affine-polygonal-tiling.ts:
```diff
@@ -89,36 +90,38 @@ export abstract class AffinePolygonalTiling<T extends AffineTile> extends Polygo
 
+      // Reminder that behavior is initialized to 'false', which takes care of case (1) of reflection over every boundary
+      // When behavior changes to 'true' (via checking a box for Snell's Law on the GUI), case (2) of Snell's Law is implemented
+      const reflected = behavior ? ray.dir.clone().rotateAround(new Vector2, -theta1 + (theta2)).multiplyScalar(2).normalize() :
```
`behavior` is a bit vague. Perhaps `snell` is a better name for this field.

ceci-fidler replied:
Done - name change made

Lael commented on angular/TilingBilliards/src/tile-billiards/affine-polygonal-tiling.ts:
```diff
@@ -89,36 +90,38 @@ export abstract class AffinePolygonalTiling<T extends AffineTile> extends Polygo
+
+    this.billiardPath = new Line(new BufferGeometry().setFromPoints(path), new LineBasicMaterial({
+      color: 0x0000ff,
+      linewidth: 4
```
For boring technical reasons, the `linewidth` attribute doesn't do anything. If you'd like this line to be thicker, you actually want a `Line2` instead of a `Line`. Take a look at the documentation if you get lost. If you don't mind that this line is one pixel wide, you can just resolve this comment.

ceci-fidler replied:
Didn't change to line2 - problem with importing linegeometry and linematerial that it needs

Lael commented on angular/TilingBilliards/src/tile-billiards/affine-polygonal-tiling.ts:
```diff
@@ -89,36 +90,38 @@ export abstract class AffinePolygonalTiling<T extends AffineTile> extends Polygo
+      linewidth: 4
+    }));
+
+    this.billiardStart = new Mesh(new CircleGeometry(0.025, 50), new MeshBasicMaterial({
```
You don't need to re-initialize this dot every time; you can just update its position.

Lael replied:
I think it would be fine for it to always be visible, so you can give it a small positive z-coordinate so the tiling gets drawn behind it. If you want to hide it sometimes, there is a `visible` attribute hiding somewhere on the `Material`.

ceci-fidler replied:
Done - also created checkbox in GUI with visible property to add/take away start point

Lael reviewed (changes requested):
Looking pretty good! Let's get this merged in, then I'll update the offset logic.

ceci-fidler committed:
More Finishing Updates

rhombusType method updated; Added functionality to give n-periodic orbit in GUI if it occurs while experimenting (The private method periodCheck is implemented such that we check if the line segment formed by the most current point in the path and the one immediately before it contains BOTH the starting point and the first path point); Updated the red dot start position such that it moves as the starting point moves, as well as the billiard path now being layered on top of it for more clarity; GUI functionality added to control whether the user wants to show the red dot start or not

Lael reviewed (approved):
LGTM
